File: leetcode/105.construct-binary-tree-from-preorder-and-inorder-traversal.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

/*
 * @lc app=leetcode id=105 lang=golang
 *
 * [105] Construct Binary Tree from Preorder and Inorder Traversal
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
/**
 * preorder = [3,9,20,15,7]
 * inorder  = [9,3,15,20,7]
 * root: 3 left_tree:[9] right_tree:[15,20,7]
 */
func buildTree(preorder []int, inorder []int) *TreeNode {
	if len(preorder) == 0 {
		return nil
	}
	root := &TreeNode{preorder[0], nil, nil}
	mid := 0
	for ; mid < len(inorder); mid++ {
		if inorder[mid] == preorder[0] {
			break
		}
	}
	root.Left = buildTree(preorder[1:len(inorder[:mid])+1], inorder[:mid])
	root.Right = buildTree(preorder[len(inorder[:mid])+1:], inorder[mid+1:])
	return root
}

// @lc code=end

File: leetcode/105.construct-binary-tree-from-preorder-and-inorder-traversal.cpp
#include <vector>
#include <unordered_map>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};
/*
 * @lc app=leetcode id=105 lang=cpp
 *
 * [105] Construct Binary Tree from Preorder and Inorder Traversal
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        if (preorder.empty()) return nullptr;
        n = preorder.size();
        for (int i = 0; i < n; i++) pos[inorder[i]] = i;
        return build(preorder, inorder, 0, n - 1, 0, n - 1);
    }

private:
    TreeNode* build(vector<int>&pre, vector<int>&in, int pl, int pr, int il, int ir) {
        if (pl > pr) return nullptr;
        TreeNode* root = new TreeNode(pre[pl]);
        int k = pos[pre[pl]];
        root->left = build(pre, in, pl + 1, pl + k - il, il, k - 1);
        root->right = build(pre, in, pl + k - il + 1, pr, k + 1, ir);
        return root;
    }
private:
    int n;
    unordered_map<int,int> pos;
};
// @lc code=end

